kernel/src/dynamic_binary_storage.rs

```diff
@@ -0,0 +1,630 @@
+                                if !metadata.setup_padding {
+                                    // Write padding header to the beginning of the new app address.
+                                    // This ensures that the linked list is not broken in the event of a
+                                    // powercycle before the app is fully written and loaded.
+
+                                    metadata.setup_padding = true;
+                                    let _ = self.write_padding_app(
+                                        metadata.new_app_length,
+                                        metadata.new_app_start_addr,
+                                    );
+                                    self.process_metadata.set(metadata);
+                                }
+                            }
+                        }
+                    };
+                    Ok(app_length)
+                }
+                Err(_err) => {
+                    // Reset the state to None because we did not find any
+                    // available address for this app.
+                    self.reset_process_loading_metadata();
+                    Err(ErrorCode::FAIL)
+                }
+            }
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    fn write_process_binary_data(
+        &self,
+        buffer: SubSliceMut<'static, u8>,
+        offset: usize,
+    ) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            State::AppWrite => {
+                let res = self.write(buffer, offset);
+                match res {
+                    Ok(()) => Ok(()),
+                    Err(e) => {
+                        // If we fail here, let us erase the app we just wrote.
+                        self.state.set(State::Fail);
+                        Err(e)
+                    }
+                }
+            }
+            _ => {
+                // We should never enter write for the rest of the conditions,
+                // so return a Busy error.
+                Err(ErrorCode::BUSY)
+            }
+        }
+    }
+
+    fn abort(&self) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            State::Setup | State::AppWrite => {
+                self.state.set(State::Abort);
+                if let Some(metadata) = self.process_metadata.get() {
+                    // Write padding header to the beginning of the new app address.
+                    // This ensures that the flash space is reclaimed for future use.
+                    match self
+                        .write_padding_app(metadata.new_app_length, metadata.new_app_start_addr)
+                    {
+                        Ok(()) => Ok(()),
+                        Err(e) => Err(e),
+                    }
+                } else {
+                    Err(ErrorCode::FAIL)
+                }
+            }
+            _ => Err(ErrorCode::BUSY),
```

Comment:
The caller needs to know whether to try the abort again, or if there was nothing to abort. Which error codes mean what for the caller?